[20.01] Fix deletion of outputs and working before job

mvdbeek opened the pull request:
Both directories may contain important files for a job, so deleting these directories may be inappropriate.
We originally implemented this for https://github.com/galaxyproject/galaxy/issues/3289

I noticed this with data managers (and potentially other OutputParameterJSONTool tools) that populate datasets in the `outputs/` directory if using `outputs_to_working_directory`. These used to be pre-populated in the job working directory, but with https://github.com/galaxyproject/galaxy/commit/ad573b215802aaaa9d09e1d2e3e2974e79295dfe they are moved to `outputs`/ (which is good), but which gets deleted during job setup.

The approach here is to copy the `working` and `outputs` directories to `_working` and `_outputs` before running the actual jobs, and if re-running the job deleting `working` and `outputs` and copying the backups back into place.
This seems to be the most robust way to make sure that resubmission by the DRM will start with proper configfiles and pre-populated outputs. The cost should be minimal as most of the time these are empty
directories.
This would also help with #9236 (comment)

bernt-matthias committed:
inputs configfile: test for filename and data_style attributes

bernt-matthias committed:
re-remove the data input

mvdbeek committed:
Only delete working/ and outputs on resubmission

mvdbeek committed:
Only delete writable files on job resubmission

mvdbeek committed:
Copy working/ and outputs/ before starting job script

This seems to be the most robust way to make sure that resubmission by
the DRM will start with proper configfiles and pre-populated outputs
(OutputParameterJSONTool).
The cost should be minimal as most of the time these are empty
directories.
This would also help with https://github.com/galaxyproject/galaxy/pull/9236#issuecomment-576050103

mvdbeek committed:
Fix unit test

jmchilton commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -19,6 +19,13 @@
+if [ -f .job_started ]; then
+    rm -rf working/ outputs/; cp -R _working working; cp -R _outputs outputs
+else
+    cp -R working _working; cp -R outputs _outputs
```
Add a random suffix to prevent collision?

natefoo replied:
Something time-derived like we do for the tmp (and home?) dirs would be preferable than random.

mvdbeek replied:
what would be the scenario for a collision ?

jmchilton replied:
Retried twice? I guess... those are cps and mvs though - maybe I didn't look at this close enough before responding...

mvdbeek replied:
The idea is that we'd always remove `working/` and `outputs/` and then copy their "backups" when the job previously created `.job_started`, so I think it should be safe ?

natefoo commented:
While this will usually be nothing, it has the potential to be very slow, especially for interrupted tools that have written GB- to TB- of intermediary/scratch files.

mvdbeek commented:
> While this will usually be nothing, it has the potential to be very slow, especially for interrupted tools that have written GB- to TB- of intermediary/scratch files.

Why is that ? It'll be an `rm -rf` just like it was before, or am I missing something ?

natefoo commented:
I see, it's like a backup of the dirs prior to any execution occurring.

mvdbeek commented:
> I see, it's like a backup of the dirs prior to any execution occurring.

yup, that's the idea!

mvdbeek commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -19,6 +19,13 @@
 [ "$GALAXY_VIRTUAL_ENV" = "None" ] && GALAXY_VIRTUAL_ENV="$_GALAXY_VIRTUAL_ENV"; _galaxy_setup_environment True
 """
+PREPARE_DIRS = """mkdir -p working outputs
+if [ -f .job_started ]; then
```
Actually this is redundant, I could just do
```suggestion
if [ -d _working]; then
```

and skip the `touch` below.

mvdbeek committed:
Simplify PREPARE_DIRS logic

mvdbeek committed:
Place configfiles in working dir if filename set

mvdbeek committed:
Test that working dir is intact on resubmission

mvdbeek committed:
Use context manager for writing param_file

mvdbeek committed:
Ignore FileNotFoundError when deleting extra files

This worked previously but logged an exception, causing further cleanup
to be skipped.

mvdbeek committed:
Things are nicer on python 3

mvdbeek committed:
Update xsd for filename attribute on configfiles
